packages/enzyme/src/ShallowWrapper.js

```diff
@@ -369,7 +369,42 @@ class ShallowWrapper {
     }
     this.single('setState', () => {
       withSetStateAllowed(() => {
-        this.instance().setState(state, callback);
+        const adapter = getAdapter(this[OPTIONS]);
+        const instance = this.instance();
+        const prevProps = instance.props;
+        const prevState = instance.state;
+        const prevContext = instance.context;
+        let shouldRender = true;
+        // This is a dirty hack but it requires to know the result of shouldComponentUpdate.
+        // When shouldComponentUpdate returns false we shouldn't call componentDidUpdate.
+        // shouldComponentUpdate is called in `instance.setState`
+        // so we replace shouldComponentUpdate to know the result and restore it later.
+        let originalShouldComponentUpdate;
+        if (
+          this[OPTIONS].lifecycleExperimental &&
+          adapter.options.enableComponentDidUpdateOnSetState &&
+          instance &&
+          typeof instance.shouldComponentUpdate === 'function'
+        ) {
+          originalShouldComponentUpdate = instance.shouldComponentUpdate;
+          instance.shouldComponentUpdate = (...args) => {
+            shouldRender = originalShouldComponentUpdate.apply(instance, args);
```

Comment:
should this function also `instance.shouldComponentUpdate = originalShouldComponentUpdate`?

I wonder if this should use `sinon` as a dependency rather than overwriting; what about things like enumerability, setters, etc?

Reply:
> I wonder if this should use sinon as a dependency rather than overwriting; what about things like enumerability, setters, etc?

Yeah, I can write this like the following if I use `sinon`.

```js
        let shouldRender = true;
        let spy;
        if (instance.shouldComponentUpdate) {
          spy = sinon.spy(instance, 'shouldComponentUpdate');
        }
        instance.setState(state, callback);
        if (spy) {
          shouldRender = spy.getCall(0).returnValue;
          spy.restore();
        }
```

Also `sinon` seems to have a better support about enumerability so I think it's worth to consider this after releasing v3. ES2015 Proxy might be an option for that.

Reply:
I'd prefer to avoid Proxy; I'm not sure it would work here anyways, and it would very tightly constrict the engines enzyme can run on.

Reply:
True. To support `< Node v6`, `sinon.spy` would be nice.

Reply:
A followup PR is always appreciated :-D

Reply:
I'll work on it 👍